src/main/java/com/mongodb/hibernate/jdbc/MongoConnectionProvider.java

```diff
@@ -0,0 +1,149 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.jdbc;
+
+import static org.bson.codecs.configuration.CodecRegistries.fromRegistries;
+import static org.hibernate.cfg.JdbcSettings.*;
+import static org.hibernate.internal.util.NullnessUtil.castNonNull;
+
+import com.mongodb.ConnectionString;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.MongoCredential;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.hibernate.cfg.ConfigurationHelper;
+import com.mongodb.hibernate.exception.ConfigurationException;
+import com.mongodb.hibernate.exception.NotYetImplementedException;
+import java.sql.Connection;
+import java.util.Map;
+import org.hibernate.engine.jdbc.connections.spi.ConnectionProvider;
+import org.hibernate.service.UnknownUnwrapTypeException;
+import org.hibernate.service.spi.Configurable;
+import org.hibernate.service.spi.Startable;
+import org.hibernate.service.spi.Stoppable;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * MongoDB dialect's customized JDBC {@link ConnectionProvider} spi implementation, whose class name is supposed to be
+ * provided as the following Hibernate property to kick off MongoDB dialect's JDBC flow:
+ *
+ * <ul>
+ *   <li>hibernate.connection.provider_class
+ * </ul>
+ *
+ * <p>The following Hibernate JDBC properties will be relied upon by Hibernate's {@link Configurable} spi mechanism:
+ *
+ * <ul>
+ *   <li>jakarta.persistence.jdbc.url
+ *   <li>jakarta.persistence.jdbc.user
+ *   <li>jakarta.persistence.jdbc.password
+ * </ul>
+ *
+ * <p><code>jakarta.persistence.jdbc.url</code> property is mandatory and it maps to MongoDB's {@link ConnectionString},
+ * in which database name must be provided to align with JDBC URL's convention. The other two JDBC properties are
+ * optional.
+ *
+ * @see ConnectionProvider
+ * @see Configurable
+ */
+public class MongoConnectionProvider implements ConnectionProvider, Configurable, Startable, Stoppable {
+
+    // non-null after configure(Map<String, Object>) method is invoked successfully
+    private @Nullable ConnectionString connectionString;
+    private @Nullable String database;
+
+    // non-null after start() method is invoked successfully
+    private @Nullable MongoClient mongoClient;
+
+    private @Nullable String user;
+    private @Nullable String password;
+
+    @Override
+    public Connection getConnection() {
+        throw new NotYetImplementedException();
+    }
+
+    @Override
+    public void closeConnection(Connection connection) {
+        throw new NotYetImplementedException();
+    }
+
+    @Override
+    public boolean supportsAggressiveRelease() {
+        return false; // won't be used in container
+    }
+
+    @Override
+    public boolean isUnwrappableAs(Class<?> unwrapType) {
```

Comment:
The two wrapping methods come from the `org.hibernate.service.spi.Wrapped` interface. As per its Javadoc (https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/service/spi/Wrapped.java):
```
Optional contract for services that wrap stuff that to which it is useful to have access.
For example, a service that maintains a javax. sql. DataSource might want to expose access to the javax. sql. DataSource or its java. sql. Connection instances.
```
`wrap` and `unwrap` mechanism is used extensively in both JDBC and Hibernate. E.g. in Hibernate's JavaType mechanism (mapping between Java type and SQL type), Java's URL type could be transformed from URL to SQL String and vice versa.

I think in this `ConnectionProvider` case, the main usage is the transforming to/from `DataSource`. But I checked all the references in Hibernate ORM and didn't find any real usage (there are minor usage in testing code but that is all).
So I guess for now a dummy implementation is enough. The above pattern is one option, but it seems the following pattern is also fine (this is the implementation in Hibernate OGM):
```
@Override
    public boolean isUnwrappableAs(Class<?> unwrapType) {
        return false;
    }

    @Override
    public <T> T unwrap(Class<T> unwrapType) {
        throw new UnknownUnwrapTypeException(unwrapType);
    }
```
if we don't intend to use these features in the first place. To my understanding, these methods should have been provided default implementation in the API definition to avoid unnecessary confusion and code boilerplate.

Testing is not needed if only dummy implementation is all we want and we never intend to use them at all.

For the other `supportsAggressiveRelease()` method, here is Gavin King's Javadoc:
```
Does this connection provider support aggressive release of JDBC connections and later re-acquisition of those connections if needed?
This is used in conjunction with org. hibernate. ConnectionReleaseMode. AFTER_STATEMENT to aggressively release JDBC connections. However, the configured ConnectionProvider must support re-acquisition of the same underlying connection for that semantic to work.
Typically, this is only true in managed environments where a container tracks connections by transaction or thread.
Note that JTA semantic depends on the fact that the underlying connection provider does support aggressive release.
Returns:
true if aggressive releasing is supported; false otherwise.
```
seems only relevant when JTA and JavaEE container usage is involved. So we simply return false and that is enough.

Reply:
To avoid confusion, I switched to the minimal dummy implementation as below:
```
@Override
    public boolean isUnwrappableAs(Class<?> unwrapType) {
        return false;
    }

    @Override
    public <T> T unwrap(Class<T> unwrapType) {
        throw new UnknownUnwrapTypeException(unwrapType);
    }
```
It has no point at all to unwrap to itself. We simply won't invoke them at all!